File: unity/ProfitableCore/Simulation/UnassignFromShipRole.cs
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/ships/unassignFromShipRole.ts. Mirrors DismissCrew's shape --
// a single, always-succeeds-unless-nothing-to-do action, no capacity/
// eligibility check needed since clearing a slot can never violate one.
public static class ShipRoleUnassigner
{
    public static UnassignShipRoleResult UnassignFromShipRole(CrewMember crewMember)
    {
        if (crewMember.ShipRole is null && crewMember.AssignedShipId is null)
        {
            return new UnassignShipRoleRejected { Reason = "crew member is not currently assigned to a ship role" };
        }

        var updatedCrewMember = new CrewMember
        {
            Id = crewMember.Id, HiredByPlayerId = crewMember.HiredByPlayerId, Tier = crewMember.Tier,
            Profession = crewMember.Profession, Status = crewMember.Status, AssignedCraftId = crewMember.AssignedCraftId,
            HiredAt = crewMember.HiredAt, LastCheckedAt = crewMember.LastCheckedAt, WageAmount = crewMember.WageAmount,
            LastPaidAt = crewMember.LastPaidAt, UnavailableUntil = crewMember.UnavailableUntil,
            ShipRole = null, AssignedShipId = null,
        };

        return new UnassignShipRoleSucceeded { UpdatedCrewMember = updatedCrewMember };
    }
}

File: src/ships/unassignFromShipRole.ts
import type { CrewMember } from "../data/types/crewMember.ts";
import type { UnassignShipRoleResult } from "../data/types/unassignShipRoleResult.ts";

// Ship Crew Roles amendment, task follow-up (ship.md's own flagged gap:
// ShipStatusScene's assignment control had no way to clear a role back to
// fully unassigned without reassigning elsewhere). Mirrors dismissCrew()'s
// shape -- a single, always-succeeds-unless-nothing-to-do action, no
// capacity/eligibility check needed since clearing a slot can never
// violate one.
export function unassignFromShipRole(crewMember: CrewMember): UnassignShipRoleResult {
  if (crewMember.shipRole == null && crewMember.assignedShipId == null) {
    return { unassigned: false, reason: "crew member is not currently assigned to a ship role" };
  }

  return {
    unassigned: true,
    updatedCrewMember: { ...crewMember, shipRole: null, assignedShipId: null },
  };
}
